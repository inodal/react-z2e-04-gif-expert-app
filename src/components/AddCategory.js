import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('')

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        // Avoid default behaviour that will submit and refresh whole view
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories(cats => [inputValue.trim(), ...cats]); // Show new first
            setInputValue('');
        }
        
    }

    return (
        <>
        <form onSubmit= { handleSubmit }>
        <p>{ inputValue }</p> 
            <input 
                type ="text"
                value = { inputValue }
                onChange = { handleInputChange }
                
            />
        </form>
            
        </>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}

