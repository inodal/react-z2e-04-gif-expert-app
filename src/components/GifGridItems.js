import React from 'react'

export const GifGridItems = ({id, title, url }) => {

    console.log(`${id}:\ntitle: ${title} \nurl: ${url}`);
    return (
        <div className="card">
            <img src={ url } alt={ title } />
            <p>{ title }</p>
        </div>
    )
}
