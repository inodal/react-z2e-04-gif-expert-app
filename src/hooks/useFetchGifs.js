import { useEffect, useState } from "react"
import { getGifs } from '../helpers/GetGif';

// All hooks start with use prefix: use{hook name} -> useFetchGifs
export const useFetchGifs = (category) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // Use hook 'useEffect' to render
    useEffect(() => {
        getGifs(category)
          .then(imgs => {
              setState({
                  data: imgs,
                  loading: false
              });
            })
    }, [category]);

    return state; // { data: [], loading: true }
}