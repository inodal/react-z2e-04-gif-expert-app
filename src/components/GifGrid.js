import React, { useState, useEffect } from 'react'
import { GifGridItems } from './GifGridItems';

export const GifGrid = ({category}) => {

    
    const [images, setImages] = useState([]);

    // Use hook 'useEffect' to render only the first time
    useEffect(() => {
        getGifs()
    }, [])

    const getGifs = async () => {
        const url = 'https://api.giphy.com/v1/gifs/search?&limit=10&q=Dragon+Ball&api_key=a6moXgoQ6vuFxhVHum2nE6CPtZdV66Uz';
        const resp = await fetch(url);
        const { data } = await resp.json();

        const gifs = data.map( img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url
            }

        });
        console.log(gifs);
        setImages(gifs);
    };

    return (
        <div>
            <h3> { category } </h3>
            {
                images.map(img => (
                    <GifGridItems 
                        key={ img.id }
                        { ...img } 
                    />
                ))
            }
        </div>
    )
}
