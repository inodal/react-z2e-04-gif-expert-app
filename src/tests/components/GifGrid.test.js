import React from 'react'
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import '@testing-library/jest-dom';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('GifGrid component test suite', () => {
    
    const category = 'Bob Esponja';

    test('should show GifGrid component properly', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category}/>);

        expect(wrapper).toMatchSnapshot();
    })

    test('should show items when images are loaded via useFetchGifs', () => {
        // Mock data
        const gifs = [{
            id: 'ABC',
            title: 'Any GIF',
            url: 'http://localhost:3000/anygif.gif'

        }, {
            id: 'ABD',
            title: 'Any JPG',
            url: 'http://localhost:3000/anyjpg.jpg'

        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        // Init component
        const wrapper = shallow(<GifGrid category={category}/>);

        // Expects
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItems').length).toBe(gifs.length);

    })

    test('should show items when images are loaded via useFetchGifs without Snapshot', () => {
        // Mock data
        const gifs = [{
            id: 'ABC',
            title: 'Any GIF',
            url: 'http://localhost:3000/anygif.gif'

        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        // Init component
        const wrapper = shallow(<GifGrid category={category}/>);

        // Expects
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItems').length).toBe(gifs.length);

    })
    
})
