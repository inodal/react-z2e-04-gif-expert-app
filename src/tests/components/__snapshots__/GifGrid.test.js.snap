// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid component test suite should show GifGrid component properly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Bob Esponja
     
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`GifGrid component test suite should show items when images are loaded via useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Bob Esponja
     
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItems
      id="ABC"
      key="ABC"
      title="Any GIF"
      url="http://localhost:3000/anygif.gif"
    />
    <GifGridItems
      id="ABD"
      key="ABD"
      title="Any JPG"
      url="http://localhost:3000/anyjpg.jpg"
    />
  </div>
</Fragment>
`;
